File: concepts/14-greedy-algorithms/python/greedy.py
"""
Greedy Algorithms in Python
"""

def activity_selection(start, finish):
    """
    Activity Selection Problem
    Select maximum non-overlapping activities
    Time: O(n log n), Space: O(1)
    """
    n = len(finish)
    # Sort by finish time
    activities = sorted(zip(finish, start))
    
    selected = [0]  # First activity always selected
    last_finish = activities[0][0]
    
    for i in range(1, n):
        if activities[i][1] >= last_finish:
            selected.append(i)
            last_finish = activities[i][0]
    
    return selected


def fractional_knapsack(weights, values, capacity):
    """
    Fractional Knapsack Problem
    Items can be taken partially
    Time: O(n log n), Space: O(n)
    """
    n = len(weights)
    items = [(values[i] / weights[i], weights[i], values[i], i) 
             for i in range(n)]
    items.sort(reverse=True)  # Sort by value/weight ratio
    
    total_value = 0
    remaining = capacity
    
    for ratio, weight, value, idx in items:
        if remaining >= weight:
            total_value += value
            remaining -= weight
        else:
            total_value += ratio * remaining
            break
    
    return total_value


def minimum_coins(coins, amount):
    """
    Minimum coins (greedy approach - works for certain coin systems)
    Time: O(n), Space: O(1)
    Note: Only works for canonical coin systems
    """
    coins.sort(reverse=True)
    count = 0
    i = 0
    
    while amount > 0 and i < len(coins):
        if coins[i] <= amount:
            num_coins = amount // coins[i]
            count += num_coins
            amount -= num_coins * coins[i]
        i += 1
    
    return count if amount == 0 else -1


def interval_scheduling(intervals):
    """
    Interval Scheduling - Maximum non-overlapping intervals
    Time: O(n log n), Space: O(1)
    """
    # Sort by end time
    intervals.sort(key=lambda x: x[1])
    
    selected = [intervals[0]]
    last_end = intervals[0][1]
    
    for interval in intervals[1:]:
        if interval[0] >= last_end:
            selected.append(interval)
            last_end = interval[1]
    
    return selected


def huffman_coding(frequencies):
    """
    Huffman Coding - Optimal prefix-free encoding
    Time: O(n log n), Space: O(n)
    """
    import heapq
    
    class Node:
        def __init__(self, char, freq, left=None, right=None):
            self.char = char
            self.freq = freq
            self.left = left
            self.right = right
        
        def __lt__(self, other):
            return self.freq < other.freq
    
    # Build priority queue
    heap = [Node(char, freq) for char, freq in frequencies.items()]
    heapq.heapify(heap)
    
    # Build Huffman tree
    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)
        merged = Node(None, left.freq + right.freq, left, right)
        heapq.heappush(heap, merged)
    
    # Generate codes
    codes = {}
    
    def generate_codes(node, code=""):
        if node.char is not None:
            codes[node.char] = code if code else "0"
        else:
            generate_codes(node.left, code + "0")
            generate_codes(node.right, code + "1")
    
    if heap:
        generate_codes(heap[0])
    
    return codes


def job_sequencing_with_deadlines(jobs):
    """
    Job Sequencing with Deadlines (Greedy)
    Schedule jobs to maximize profit before deadlines
    Each job: (job_id, profit, deadline)
    Time: O(n²), Space: O(n)
    """
    # Sort by profit (descending)
    jobs.sort(key=lambda x: x[1], reverse=True)
    
    # Find maximum deadline
    max_deadline = max(job[2] for job in jobs)
    
    # Initialize result array
    result = [None] * (max_deadline + 1)
    total_profit = 0
    
    for job in jobs:
        job_id, profit, deadline = job
        # Find latest free slot before deadline
        for j in range(deadline, 0, -1):
            if result[j] is None:
                result[j] = job_id
                total_profit += profit
                break
    
    # Get scheduled jobs (excluding None)
    scheduled = [job for job in result if job is not None]
    return scheduled, total_profit


def weighted_job_scheduling(jobs):
    """
    Weighted Job Scheduling (DP approach)
    Each job: (start, end, profit)
    Find maximum profit subset of non-overlapping jobs
    Time: O(n²), Space: O(n)
    """
    # Sort by end time
    jobs.sort(key=lambda x: x[1])
    n = len(jobs)
    
    # DP[i] = maximum profit using jobs[0..i]
    dp = [0] * n
    dp[0] = jobs[0][2]
    
    for i in range(1, n):
        # Profit including current job
        profit_including = jobs[i][2]
        
        # Find last job that doesn't conflict
        for j in range(i - 1, -1, -1):
            if jobs[j][1] <= jobs[i][0]:
                profit_including += dp[j]
                break
        
        # Maximum of including or excluding current job
        dp[i] = max(profit_including, dp[i - 1])
    
    return dp[n - 1]


def job_scheduling_earliest_deadline(jobs):
    """
    Job Scheduling - Earliest Deadline First (EDF)
    Schedule jobs to minimize maximum lateness
    Each job: (job_id, processing_time, deadline)
    Time: O(n log n), Space: O(n)
    """
    # Sort by deadline
    jobs.sort(key=lambda x: x[2])
    
    schedule = []
    current_time = 0
    max_lateness = 0
    
    for job in jobs:
        job_id, processing_time, deadline = job
        finish_time = current_time + processing_time
        lateness = max(0, finish_time - deadline)
        max_lateness = max(max_lateness, lateness)
        
        schedule.append((job_id, current_time, finish_time, lateness))
        current_time = finish_time
    
    return schedule, max_lateness


# Example usage
if __name__ == "__main__":
    # Activity Selection
    start = [1, 3, 0, 5, 8, 5]
    finish = [2, 4, 6, 7, 9, 9]
    print("Activity Selection:", activity_selection(start, finish))
    
    # Fractional Knapsack
    weights = [10, 20, 30]
    values = [60, 100, 120]
    capacity = 50
    print("Fractional Knapsack value:", fractional_knapsack(weights, values, capacity))
    
    # Minimum Coins
    coins = [1, 5, 10, 25]
    amount = 67
    print("Minimum coins:", minimum_coins(coins, amount))
    
    # Interval Scheduling
    intervals = [(1, 3), (2, 5), (4, 7), (6, 9), (8, 10)]
    print("Interval Scheduling:", interval_scheduling(intervals))
    
    # Huffman Coding
    frequencies = {'a': 5, 'b': 9, 'c': 12, 'd': 13, 'e': 16, 'f': 45}
    print("Huffman Codes:", huffman_coding(frequencies))
    
    # Job Sequencing with Deadlines
    print("\n--- Job Sequencing with Deadlines ---")
    jobs = [
        ('a', 20, 2), ('b', 15, 2), ('c', 10, 1),
        ('d', 5, 3), ('e', 1, 3)
    ]
    scheduled, profit = job_sequencing_with_deadlines(jobs)
    print(f"Scheduled jobs: {scheduled}, Total profit: {profit}")
    
    # Weighted Job Scheduling
    print("\n--- Weighted Job Scheduling ---")
    weighted_jobs = [
        (1, 3, 5), (2, 5, 6), (4, 6, 5),
        (6, 7, 4), (5, 8, 11), (7, 9, 2)
    ]
    max_profit = weighted_job_scheduling(weighted_jobs)
    print(f"Maximum profit: {max_profit}")
    
    # Earliest Deadline First
    print("\n--- Earliest Deadline First ---")
    edf_jobs = [
        ('J1', 2, 6), ('J2', 1, 4), ('J3', 3, 5),
        ('J4', 2, 7), ('J5', 1, 3)
    ]
    schedule, lateness = job_scheduling_earliest_deadline(edf_jobs)
    print(f"Schedule: {schedule}")
    print(f"Maximum lateness: {lateness}")

File: concepts/14-greedy-algorithms/node/greedy.js
/**
 * Greedy Algorithms in Node.js
 */

/**
 * Activity Selection Problem
 * Select maximum non-overlapping activities
 * Time: O(n log n), Space: O(1)
 */
function activitySelection(start, finish) {
    const n = start.length;
    const activities = finish.map((f, i) => [f, start[i], i]).sort((a, b) => a[0] - b[0]);

    const selected = [0];
    let lastFinish = activities[0][0];

    for (let i = 1; i < n; i++) {
        if (activities[i][1] >= lastFinish) {
            selected.push(activities[i][2]);
            lastFinish = activities[i][0];
        }
    }

    return selected;
}

/**
 * Fractional Knapsack Problem
 * Items can be taken partially
 * Time: O(n log n), Space: O(n)
 */
function fractionalKnapsack(weights, values, capacity) {
    const n = weights.length;
    const items = weights.map((w, i) => [values[i] / w, w, values[i], i])
        .sort((a, b) => b[0] - a[0]);

    let totalValue = 0;
    let remaining = capacity;

    for (const [ratio, weight, value] of items) {
        if (remaining >= weight) {
            totalValue += value;
            remaining -= weight;
        } else {
            totalValue += ratio * remaining;
            break;
        }
    }

    return totalValue;
}

/**
 * Minimum coins (greedy approach - works for certain coin systems)
 * Time: O(n), Space: O(1)
 */
function minimumCoins(coins, amount) {
    coins.sort((a, b) => b - a);
    let count = 0;
    let i = 0;

    while (amount > 0 && i < coins.length) {
        if (coins[i] <= amount) {
            const numCoins = Math.floor(amount / coins[i]);
            count += numCoins;
            amount -= numCoins * coins[i];
        }
        i++;
    }

    return amount === 0 ? count : -1;
}

/**
 * Interval Scheduling - Maximum non-overlapping intervals
 * Time: O(n log n), Space: O(1)
 */
function intervalScheduling(intervals) {
    intervals.sort((a, b) => a[1] - b[1]);

    const selected = [intervals[0]];
    let lastEnd = intervals[0][1];

    for (let i = 1; i < intervals.length; i++) {
        if (intervals[i][0] >= lastEnd) {
            selected.push(intervals[i]);
            lastEnd = intervals[i][1];
        }
    }

    return selected;
}

/**
 * Huffman Coding - Optimal prefix-free encoding
 * Time: O(n log n), Space: O(n)
 */
function huffmanCoding(frequencies) {
    class Node {
        constructor(char, freq, left = null, right = null) {
            this.char = char;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }
    }

    const heap = Object.entries(frequencies)
        .map(([char, freq]) => new Node(char, freq))
        .sort((a, b) => a.freq - b.freq);

    while (heap.length > 1) {
        const left = heap.shift();
        const right = heap.shift();
        const merged = new Node(null, left.freq + right.freq, left, right);
        heap.push(merged);
        heap.sort((a, b) => a.freq - b.freq);
    }

    const codes = {};

    function generateCodes(node, code = "") {
        if (node.char !== null) {
            codes[node.char] = code || "0";
        } else {
            generateCodes(node.left, code + "0");
            generateCodes(node.right, code + "1");
        }
    }

    if (heap.length > 0) {
        generateCodes(heap[0]);
    }

    return codes;
}

/**
 * Job Sequencing with Deadlines (Greedy)
 * Schedule jobs to maximize profit before deadlines
 * Each job: [job_id, profit, deadline]
 * Time: O(n²), Space: O(n)
 */
function jobSequencingWithDeadlines(jobs) {
    // Sort by profit (descending)
    jobs.sort((a, b) => b[1] - a[1]);

    // Find maximum deadline
    const maxDeadline = Math.max(...jobs.map(job => job[2]));

    // Initialize result array
    const result = new Array(maxDeadline + 1).fill(null);
    let totalProfit = 0;

    for (const job of jobs) {
        const [jobId, profit, deadline] = job;
        // Find latest free slot before deadline
        for (let j = deadline; j > 0; j--) {
            if (result[j] === null) {
                result[j] = jobId;
                totalProfit += profit;
                break;
            }
        }
    }

    // Get scheduled jobs (excluding null)
    const scheduled = result.filter(job => job !== null);
    return { scheduled, totalProfit };
}

/**
 * Weighted Job Scheduling (DP approach)
 * Each job: [start, end, profit]
 * Find maximum profit subset of non-overlapping jobs
 * Time: O(n²), Space: O(n)
 */
function weightedJobScheduling(jobs) {
    // Sort by end time
    jobs.sort((a, b) => a[1] - b[1]);
    const n = jobs.length;

    // DP[i] = maximum profit using jobs[0..i]
    const dp = new Array(n).fill(0);
    dp[0] = jobs[0][2];

    for (let i = 1; i < n; i++) {
        // Profit including current job
        let profitIncluding = jobs[i][2];

        // Find last job that doesn't conflict
        for (let j = i - 1; j >= 0; j--) {
            if (jobs[j][1] <= jobs[i][0]) {
                profitIncluding += dp[j];
                break;
            }
        }

        // Maximum of including or excluding current job
        dp[i] = Math.max(profitIncluding, dp[i - 1]);
    }

    return dp[n - 1];
}

/**
 * Job Scheduling - Earliest Deadline First (EDF)
 * Schedule jobs to minimize maximum lateness
 * Each job: [job_id, processing_time, deadline]
 * Time: O(n log n), Space: O(n)
 */
function jobSchedulingEarliestDeadline(jobs) {
    // Sort by deadline
    jobs.sort((a, b) => a[2] - b[2]);

    const schedule = [];
    let currentTime = 0;
    let maxLateness = 0;

    for (const job of jobs) {
        const [jobId, processingTime, deadline] = job;
        const finishTime = currentTime + processingTime;
        const lateness = Math.max(0, finishTime - deadline);
        maxLateness = Math.max(maxLateness, lateness);

        schedule.push([jobId, currentTime, finishTime, lateness]);
        currentTime = finishTime;
    }

    return { schedule, maxLateness };
}

// Example usage
if (require.main === module) {
    const start = [1, 3, 0, 5, 8, 5];
    const finish = [2, 4, 6, 7, 9, 9];
    console.log("Activity Selection:", activitySelection(start, finish));

    const weights = [10, 20, 30];
    const values = [60, 100, 120];
    const capacity = 50;
    console.log("Fractional Knapsack value:", fractionalKnapsack(weights, values, capacity));

    const coins = [1, 5, 10, 25];
    const amount = 67;
    console.log("Minimum coins:", minimumCoins(coins, amount));

    const intervals = [[1, 3], [2, 5], [4, 7], [6, 9], [8, 10]];
    console.log("Interval Scheduling:", intervalScheduling(intervals));

    const frequencies = { 'a': 5, 'b': 9, 'c': 12, 'd': 13, 'e': 16, 'f': 45 };
    console.log("Huffman Codes:", huffmanCoding(frequencies));

    // Job Sequencing with Deadlines
    console.log("\n--- Job Sequencing with Deadlines ---");
    const jobs = [
        ['a', 20, 2], ['b', 15, 2], ['c', 10, 1],
        ['d', 5, 3], ['e', 1, 3]
    ];
    const { scheduled, totalProfit } = jobSequencingWithDeadlines(jobs);
    console.log(`Scheduled jobs: [${scheduled.join(', ')}], Total profit: ${totalProfit}`);

    // Weighted Job Scheduling
    console.log("\n--- Weighted Job Scheduling ---");
    const weightedJobs = [
        [1, 3, 5], [2, 5, 6], [4, 6, 5],
        [6, 7, 4], [5, 8, 11], [7, 9, 2]
    ];
    const maxProfit = weightedJobScheduling(weightedJobs);
    console.log(`Maximum profit: ${maxProfit}`);

    // Earliest Deadline First
    console.log("\n--- Earliest Deadline First ---");
    const edfJobs = [
        ['J1', 2, 6], ['J2', 1, 4], ['J3', 3, 5],
        ['J4', 2, 7], ['J5', 1, 3]
    ];
    const { schedule, maxLateness } = jobSchedulingEarliestDeadline(edfJobs);
    console.log("Schedule:", schedule);
    console.log(`Maximum lateness: ${maxLateness}`);
}

module.exports = {
    activitySelection,
    fractionalKnapsack,
    minimumCoins,
    intervalScheduling,
    huffmanCoding,
    jobSequencingWithDeadlines,
    weightedJobScheduling,
    jobSchedulingEarliestDeadline
};

